### PR TITLE
Performance fixes and fix manual sort

### LINKED ISSUE #7186: Favorites and channels lists are reversed
I just opened the app to send a message, and I noticed that items in my favorites and channels are reversed. The one I usually have at the top is now at the bottom of the section, and so on. 

It still looks fine on the desktop app, so I have to assume this is a mobile version bug.

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Fix performance issues introduced by #7170
Also fix the manual sort being reversed, introduced on the same PR

Also, piggybacked some minor performance improvements.

#### Ticket Link
Fix #7186 

#### Release Note
```release-note
Fix performance issue on team switch.
Minor performance improvements on team switch.
Fix channels being reversed on manually sorted categories.
```

### CODE REVIEW

File: app/utils/categories.ts [hunk 1]
```diff
@@ -42,15 +42,18 @@ export const filterAutoclosedDMs = (
...
+    }, new Map<string, number>());
     const getLastViewedAt = (cwm: ChannelWithMyChannel) => {
         // The server only ever sets the last_viewed_at to the time of the last post in channel, so we may need
         // to use the preferences added for the previous version of autoclosing DMs.
         const id = cwm.channel.id;
         return Math.max(
             cwm.myChannel.lastViewedAt,
-            getPreferenceValue<number>(preferences, Preferences.CATEGORIES.CHANNEL_APPROXIMATE_VIEW_TIME, id, 0),
-            getPreferenceValue<number>(preferences, Preferences.CATEGORIES.CHANNEL_OPEN_TIME, id, 0),
+            prefMap.get(id) || 0,
```
Comment (larkox) [inline app/utils/categories.ts:56]: This is where the real issue was. `getPreferenceValue` performs a find over all the preferences, which in this case they are a lot. That is done every channel, and potentially several times.

File: app/utils/categories.ts [hunk 2]
```diff
@@ -178,7 +181,7 @@ export const sortChannels = (sorting: CategorySorting, channelsWithMyChannel: Ch
         }).map((cwm) => cwm.channel);
     } else if (sorting === 'manual') {
         return channelsWithMyChannel.sort((cwmA, cwmB) => {
-            return cwmB.sortOrder - cwmA.sortOrder;
+            return cwmA.sortOrder - cwmB.sortOrder;
```
Comment (larkox) [inline app/utils/categories.ts:184]: This solves the order issue.

File: app/screens/home/channel_list/index.ts [hunk 3]
```diff
@@ -21,11 +21,22 @@ const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
         switchMap((lcs) => (lcs ? of$(lcs.IsLicensed === 'true') : of$(false))),
     );
 
+    const teamsCount = queryMyTeams(database).observeCount(false);
+
     return {
         isCRTEnabled: observeIsCRTEnabled(database),
-        teamsCount: queryMyTeams(database).observeCount(false),
-        channelsCount: observeCurrentTeamId(database).pipe(
+        hasTeams: teamsCount.pipe(
```
Comment (larkox) [inline app/screens/home/channel_list/index.ts:28]: These changes help with minor unneeded re-renders on team switch due to the number of channels changing, when our real interest is whether we have more than one or not.

File: app/screens/home/channel_list/categories_list/header/index.ts [hunk 4]
```diff
@@ -25,6 +25,7 @@ const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
 
     const canJoinChannels = combineLatest([currentUser, team]).pipe(
         switchMap(([u, t]) => observePermissionForTeam(database, t, u, Permissions.JOIN_PUBLIC_CHANNELS, true)),
+        distinctUntilChanged(),
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/header/index.ts:28]: These changes help with the categories header not being re-rendered more than needed.

observePermissionForTeam already check for distincUntilChanged, but the team changing recreates the observer, so we need an extra distinctUntilChanged here.

File: app/screens/home/channel_list/categories_list/categories/index.ts [hunk 5]
```diff
@@ -23,12 +23,23 @@ const enhanced = withObservables(
...
+        const unreadsOnTopUserPreference = querySidebarPreferences(database, Preferences.CHANNEL_SIDEBAR_GROUP_UNREADS).
             observeWithColumns(['value']).
             pipe(
-                switchMap((prefs: PreferenceModel[]) => of$(getSidebarPreferenceAsBool(prefs, Preferences.CHANNEL_SIDEBAR_GROUP_UNREADS))),
+                switchMap((prefs: PreferenceModel[]) => of$(getPreferenceValue<string>(prefs, Preferences.CATEGORIES.SIDEBAR_SETTINGS, Preferences.CHANNEL_SIDEBAR_GROUP_UNREADS))),
             );
 
+        const unreadsOnTopServerPreference = observeConfigBooleanValue(database, 'ExperimentalGroupUnreadChannels');
+
+        const unreadsOnTop = combineLatest([unreadsOnTopUserPreference, unreadsOnTopServerPreference]).pipe(
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/index.ts:34]: This adds the `unreadsOnTop` functionality based on the server preference, to add parity with web.

File: app/screens/home/channel_list/categories_list/categories/body/index.ts [hunk 6]
```diff
@@ -57,7 +57,8 @@ const observeCategoryChannels = (category: CategoryModel) => {
 };
 
 const enhanced = withObservables([], ({category, currentUserId, database, isTablet, locale}: EnhanceProps) => {
-    const channelsWithMyChannel = observeCategoryChannels(category);
+    const categoryMyChannels = category.myChannels.observeWithColumns(['last_post_at', 'is_unread']);
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/body/index.ts:60]: This same observer is used here and in observeCategoryChannels. This way we only create it once, and we don't have to use the map that appears later.

**Review by larkox — COMMENTED** [review]
Some comments about the changes.

File: app/screens/home/channel_list/categories_list/categories/index.ts [hunk 7]
```diff
@@ -3,14 +3,14 @@
 
 import {withDatabase} from '@nozbe/watermelondb/DatabaseProvider';
 import withObservables from '@nozbe/with-observables';
-import {of as of$} from 'rxjs';
+import {of as of$, combineLatest} from 'rxjs';
```
Comment (enahum) [inline app/screens/home/channel_list/categories_list/categories/index.ts:6]: Use combineLatestWith instead, check the docs to understand the difference, but mainly it wmits once all values are set at least once

File: app/screens/home/channel_list/categories_list/categories/index.ts (hunk 5 above)
Comment (enahum) [reply app/screens/home/channel_list/categories_list/categories/index.ts:34]: Why was this added? Does it need the experimental config? I'm almost sure that got removed

**Review by enahum — COMMENTED** [review]
Some comments before approving.

Btw thanks for looking into this

File: app/screens/home/channel_list/categories_list/categories/index.ts (hunk 5 above)
Comment (larkox) [reply app/screens/home/channel_list/categories_list/categories/index.ts:34]: This came up in the grooming meeting. Maybe @hmhealey has better context.

**Review by enahum — APPROVED** [review]
The one thing pending is the experimental config which I want to understand before shipping..

**Comment by larkox** [comment]
@enahum I reverted the change, so we can merge this, and I will add a new PR just for that later.

**Pull request merged**